### PR TITLE
WIP attempt at fixing recovery-straight-path issues

### PR DESCRIPTION
**Pull request by SteveMacenski**
Follow up on #1999 and @naiveHobo 's #2009 

@naiveHobo can you take a look over this and see if it seems reasonable to you?
- Added clearability to layer API (need to add migration docs for this and explain what its for)
- Greatly simplified the clear costmap service code
- Updated the `current_` to be active: so having run after a clear request. If I did this correctly, it is true after initialization,  processing and an update, but should be false only after resetting the layers or clearing areas for update.
-  controller / planner servers check if active and block until active after a clear

@pfedom please test and see if this fixes your issue

- [x] lint (which I'm sure there's issues with)
- [x] have users test that this fixes issues
- [x] a couple reviews to make sure I didn't miss a subtle detail
- [x] documentation on API / `current_` change, migration guides https://github.com/ros-planning/navigation.ros.org/pull/114
- [ ] ~Update other community plugins (STVL, NPVL)~ new ticket #2089 
- [x] double check all the `current_` changes are set where they should be and to the correct values

### CODE REVIEW

**Comment by DLu** [comment]
Still working on reading through the code, but a couple points:
 * In Navigation 1, `isCurrent()` is not used within `costmap_2d`, but crucially is used by [`move_base`](https://github.com/ros-planning/navigation/blob/a9bc9c4c35a55390963db1357926ec461fcff24c/move_base/src/move_base.cpp#L827) to stop the robot when its not current, which most likely means you dropped sensor data. [Full Overview](https://github.com/ros-planning/navigation/search?q=isCurrent%7Ccurrent)
 * The problem of planning immediately after clearing was addressed in [this PR](https://github.com/ros-planning/navigation/pull/838), which added a flag to force what @naiveHobo was doing in #2009. However, I could easily believe that this PR was merged *after* the ROS2 Navigation2 code was "forked". 
 * I have no good explanation for why this problem was not encountered more in Nav1, except that usually the costmap update frequency was usually higher than the planner frequency.

**Comment by SteveMacenski** [comment]
On your first point: Ah ok, that makes sense. I think that's relatively analogous to how we're using it in Nav2 in the planner / controller servers. Now we're making sure that they're valid from a clearing operation. We could update it in the obstacle / voxel layers to also check for recently updated (to fix clearing issue) **and** also current in terms of measurements. Right now, its just on the updated after an invalidation operation like clearing / resetting. 

On 2: Do you think that's the better way to do this? I thought that updating the map in this callback would be messing up the regular update frequency which I didn't think would be great. If you don't think that's a problem, then all of this is overkill and we should just do that. We'd need to add mutex locks on `updateMap` so that we don't have 2 updates happening at the same time.

**Review by naiveHobo — APPROVED** [review]

**Comment by naiveHobo** [comment]
Really happy with the draft, I've tested your changes and it resolved #1999. I think this is an overall better solution than explicitly force updating the costmap after a clear costmap service is called. A few uncrustify errors but otherwise everything looks great to me!

**Comment by SteveMacenski** [comment]
@DLu does this seem good to you if we change current to mean dropping data & not after clearing operation?

File: nav2_controller/src/nav2_controller.cpp [hunk 1]
```diff
@@ -297,6 +297,12 @@ void ControllerServer::computeControl()
         return;
       }
 
+      // Don't compute a trajectory until costmap is valid (after clear costmap)
+      rclcpp::Rate r(100);
+      while (!costmap_ros_->isCurrent()) {
+        r.sleep();
```
Comment (DLu) [inline nav2_controller/src/nav2_controller.cpp:303]: I'm not generally familiar with the entire threading model for Nav2 yet, but I don't like this spinning until its ready. Is there a reason for not publishing zero velocity and just trying again next time?

File: nav2_costmap_2d/include/nav2_costmap_2d/layer.hpp [hunk 2]
```diff
@@ -68,6 +68,7 @@ class Layer  // TODO(mjeronimo): public nav2_util::LifecycleHelperInterface
   virtual void deactivate() {} /** @brief Stop publishers. */
   virtual void activate() {}   /** @brief Restart publishers if they've been stopped. */
   virtual void reset() = 0;
+  virtual bool isClearable() = 0;
```
Comment (DLu) [inline nav2_costmap_2d/include/nav2_costmap_2d/layer.hpp:71]: This seems a little outside the scope of the original PR, but I think the burden of figuring out which layers to clear should stay with the recovery behavior rather than making a flag for it in the layer interface itself.

Comment (SteveMacenski) [reply nav2_costmap_2d/include/nav2_costmap_2d/layer.hpp:71]: Why? Is this not a property of the layer itself whether it is clearable?

File: nav2_controller/src/nav2_controller.cpp (hunk 1 above)
Comment (SteveMacenski) [reply nav2_controller/src/nav2_controller.cpp:303]: Publishing zero could have jerky motion and that analog doesn't really work for the planner.

File: nav2_costmap_2d/include/nav2_costmap_2d/layer.hpp (hunk 2 above)
Comment (DLu) [reply nav2_costmap_2d/include/nav2_costmap_2d/layer.hpp:71]: I'd rather put the burden on the person configuring the system rather than force every layer-implementer to code for every possible recovery behavior.

**Comment by SteveMacenski** [comment]
@pfedom did you test that this resolves your issue?

File: nav2_controller/src/nav2_controller.cpp (hunk 1 above)
Comment (SteveMacenski) [reply nav2_controller/src/nav2_controller.cpp:303]: Either way, the controller server just spins at the requested spin rate, so waiting here or skipping and updating the next update cycle (without sending a zero command) really wouldn't have much of a difference. It just helps so that if you're spinning at 20 hz and there's an update, rather than getting 19/20 out, you might still get 20/20 out but one of them had a little jitter to it. I'd think we prefer jitter over missing update - but that's open for argument.

What do you think?

File: nav2_costmap_2d/include/nav2_costmap_2d/layer.hpp (hunk 2 above)
Comment (SteveMacenski) [reply nav2_costmap_2d/include/nav2_costmap_2d/layer.hpp:71]: @DLu @mikeferguson  what are your thoughts here?

tl;dr, changed to add `isClearable()` to layer API so that its a property of the costmap layer rather than a parameter

**Comment by SteveMacenski** [comment]
https://github.com/ros-planning/navigation2/pull/2044/commits/acd15a7d9630d9f91f682e9820e9dfed8eaeb753 changes things so that we  can properly handle @DLu 's explanation of the original `current_` use. It now allows for layers that are `current_` to false due to stale / no measurements to work alongside the reset case. I think this is the best of all worlds:
- Allows for layers like the range layer that are `current_` to false when expected update rates are not respected and does not override it when updated
- Allows for layers when `reset()` to return false on `current_` until at least 1 successful update cycle is had, so that we don't say that we're current, when we're really not because we were just reset and haven't yet updated anything
- Allows for the intended check to block processing algorithms until all the layers are valid with some current data sources

**Comment by pfedom** [comment]
Sorry for the late answer. I had many things to do and I can not get nav2:main going at the moment. 
I will post my errors at slack. 
@gramss will test it for me and will give you feedback.

**Comment by gramss** [comment]
After locally rebasing this branch on current main to get it working with current ROS2 Rolling, I can say that this fix is pretty satisfying! 

I cannot get the robot to reproduce the error (plan straight lines through walls). Out of the user perspective the robot only drives towards a goal behind a wall if it is within a given threshold of the costmap.
This works really robust on my machine. Cannot complain at all.. :)

I also tried some corner cases. Multiple goals in the V shaped corner outside of the map at the "northest" point. After 10 Goals intentional inside the outside Edge of the threshold to accept a goal but not already inside the threshold of "goal reached", I still got the robot pretty close to the wall. It only get stock a little bit, but was able to fully recover and tackle a new normal goals just fine.. :)

So, the original error should be cleared with this. What I couldn't test is creating a map with open doors and then closing the door and see what happens. This was the original error case where our agv drove straight towards the next wall.. I think it would not drive straight against the wall but would try it until it reaches the recovery-retry threshold. But this is a thing for another day.

So nothing to complain on my side!

**Comment by SteveMacenski** [comment]
https://github.com/ros-planning/navigation2/pull/2090 supersedes
